Login crash

- Occurs when the user logs out then tries signing back in
- cc: @orestesgaolin I am noticing this on the latest dev, could it be related to #653 ?
```
flutter: Converting object to an encodable object failed: Instance of 'UserData'
flutter: #0      _JsonStringifier.writeObject (dart:convert/json.dart:659:7)
flutter: #1      _JsonStringStringifier.printOn (dart:convert/json.dart:846:17)
flutter: #2      _JsonStringStringifier.stringify (dart:convert/json.dart:831:5)
flutter: #3      JsonEncoder.convert (dart:convert/json.dart:260:30)
flutter: #4      JsonCodec.encode (dart:convert/json.dart:168:45)
flutter: #5      jsonEncode (dart:convert/json.dart:82:10)
flutter: #6      AuthenticationServiceCentralized.loginUser (package:junto_beta_mobile/backend/services/auth_service.dart:38:41)
flutter: <asynchronous suspension>
flutter: #7      AuthRepo.loginUser (package:junto_beta_mobile/backend/repositories/auth_repo.dart:72:48)
flutter: #8      _SignInState._handleSignIn (package:junto_beta_mobile/screens/welcome/sign_in.dart:71:12)
flutter: <asynchronous suspension>
flutter: #9      _SignInState.build.<anonymous closure> (package:junto_beta_mobile/screens/welcome/sign_in.dart:150:21)
flutter: #10     _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:779:19)
flutter: #11     _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:862:36)
flutter: #12     GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
flutter: #13     TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:504:11)
flutter: #14     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:282:5)
flutter: #15     BaseTapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:254:7)
flutter: #16     GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:156:27)
flutter: #17     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:222:20)
flutter: #18     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:198:22)
flutter: #19     GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:156:7)
flutter: #20     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:102:7)
flutter: #21     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:86:7)
flutter: #22     _rootRunUnary (dart:async/zone.dart:1196:13)
flutter: #23     _CustomZone.runUnary (dart:async/zone.dart:1085:19)
flutter: #24     _CustomZone.runUnaryGuarded (dart:async/zone.dart:987:7)
flutter: #25     _invoke1 (dart:ui/hooks.dart:275:10)
flutter: #26     _dispatchPointerDataPacket (dart:ui/hooks.dart:184:5)
flutter:
flutter: [ERROR] [PrintLogger.logException]: Caught error: Converting object to an encodable object failed: Instance of 'UserData'
flutter: [DEBUG] [PrintLogger.logException]: In dev mode. Not sending report to Sentry.io.
flutter: [DEBUG] [_RouteEntry.handlePush]: push null from /

```

--- a/lib/backend/backend.dart
+++ b/lib/backend/backend.dart
@@ -65,7 +65,7 @@ class Backend {
       );
       return Backend._(
         searchRepo: SearchRepo(searchService),
-        authRepo: AuthRepo(authService, userRepo),
+        authRepo: AuthRepo(authService, userRepo, expressionService),
         userRepo: userRepo,
         collectiveProvider: CollectiveProviderCentralized(client),
         groupsProvider: GroupRepo(groupService, userService),
@@ -87,7 +87,7 @@ class Backend {
     final GroupService groupService = MockSphere();
     final SearchService searchService = MockSearch();
     return Backend._(
-        authRepo: AuthRepo(authService, null),
+        authRepo: AuthRepo(authService, null, expressionService),
         userRepo: UserRepo(userService, null, null),
         collectiveProvider: null,
         groupsProvider: GroupRepo(groupService, userService),

--- a/lib/backend/repositories/auth_repo.dart
+++ b/lib/backend/repositories/auth_repo.dart
@@ -10,9 +10,10 @@ import 'package:junto_beta_mobile/hive_keys.dart';
 import 'package:junto_beta_mobile/models/models.dart';
 
 class AuthRepo {
-  const AuthRepo(this.authService, this.userRepo);
+  const AuthRepo(this.authService, this.userRepo, this.expressionService);
 
   final AuthenticationService authService;
+  final ExpressionService expressionService;
   final UserRepo userRepo;
 
   Future<bool> isLoggedIn() async {
@@ -45,7 +46,8 @@ class AuthRepo {
   }
 
   /// Registers a user on the server and creates their profile.
-  Future<UserData> registerUser(UserAuthRegistrationDetails details) async {
+  Future<UserData> registerUser(
+      UserAuthRegistrationDetails details, File profilePicture) async {
     logger.logDebug('Registering user');
     final UserData _data = await authService.registerUser(details);
     await authService.loginUser(
@@ -54,6 +56,24 @@ class AuthRepo {
         password: details.password,
       ),
     );
+
+    if (profilePicture != null) {
+      final key =
+          await expressionService.createPhoto(false, '.png', profilePicture);
+
+      //TODO: WHAT IS THAT? WHY ARE WE CREATING MAPS MANUALLY?
+      final _profilePictureKeys = <String, dynamic>{
+        'profile_picture': <Map<String, dynamic>>[
+          <String, dynamic>{'index': 0, 'key': key},
+        ]
+      };
+      // update user with profile photos
+      await userRepo.updateUser(
+        _profilePictureKeys,
+        _data.user.address,
+      );
+    }
+
     final box = await Hive.openLazyBox(HiveBoxes.kAppBox, encryptionKey: key);
     await box.put(HiveKeys.kisLoggedIn, true);
     await box.put(HiveKeys.kUserId, _data.user.address);

--- a/lib/backend/services/auth_service.dart
+++ b/lib/backend/services/auth_service.dart
@@ -30,12 +30,15 @@ class AuthenticationServiceCentralized implements AuthenticationService {
       },
     );
     if (response.statusCode == 200) {
+      logger.logInfo('User logged in');
       final String authorization = response.headers['authorization'];
       final box = await Hive.openLazyBox(HiveBoxes.kAppBox, encryptionKey: key);
       await box.put(HiveKeys.kAuth, authorization);
       final userData = JuntoHttp.handleResponse(response);
       final user = UserData.fromMap(userData);
-      await box.put(HiveKeys.kUserData, jsonEncode(user));
+      final userMap = jsonEncode(userData);
+      logger.logInfo('Storing user data');
+      await box.put(HiveKeys.kUserData, userMap);
       return user;
     } else {
       final Map<String, dynamic> errorResponse =

--- a/lib/backend/services/auth_service.dart
+++ b/lib/backend/services/auth_service.dart
@@ -81,8 +81,7 @@ class AuthenticationServiceCentralized implements AuthenticationService {
     logger.logDebug(_body.toString());
     final http.Response response =
         await client.postWithoutEncoding('/auth/register', body: _body);
-    logger.logDebug(response.body);
-    print(response.statusCode);
+    logger.logDebug('${response.statusCode} ${response.body}');
     if (response.statusCode == 310) {
       print('yeo');
       return 'follow the white rabbit';

--- a/lib/backend/services/expression_provider.dart
+++ b/lib/backend/services/expression_provider.dart
@@ -38,50 +38,55 @@ class ExpressionServiceCentralized implements ExpressionService {
 
   @override
   Future<String> createPhoto(bool isPrivate, String fileType, File file) async {
-    String _serverUrl;
-    if (isPrivate) {
-      _serverUrl = '/auth/s3?private=true';
-    } else if (isPrivate == false) {
-      _serverUrl = '/auth/s3?private=false';
-    }
-    final contentLength = file.lengthSync();
+    try {
+      logger.logDebug('Creating a photo');
+      String _serverUrl;
+      if (isPrivate) {
+        _serverUrl = '/auth/s3?private=true';
+      } else if (isPrivate == false) {
+        _serverUrl = '/auth/s3?private=false';
+      }
+      final contentLength = file.lengthSync();
 
-    // denote file type and get url, headers, and key of s3 bucket
-    final http.Response _serverResponse = await client.postWithoutEncoding(
-      _serverUrl,
-      body: {'content_type': fileType, 'content_length': contentLength},
-    );
-    logger.logDebug(_serverResponse.body);
-    logger.logDebug(_serverResponse.statusCode.toString());
-
-    // parse response
-    final Map<String, dynamic> parseData =
-        JuntoHttp.handleResponse(_serverResponse);
-
-    // seralize into new headers
-    final Map<String, String> newHeaders = <String, String>{
-      'x-amz-acl': parseData['headers']['x-amz-acl'],
-      'x-amz-meta-user_id': parseData['headers']['x-amz-meta-user_id']
-    };
-
-    // turn file into bytes
-    final Uint8List fileAsBytes = file.readAsBytesSync();
-
-    // send put request to s3 bucket with url, new headers, and file as bytes
-    final http.Response _serverResponseTwo = await http.put(
-      parseData['signed_url'],
-      headers: newHeaders,
-      body: fileAsBytes,
-    );
-
-    // if successful, return the key for next steps
-    if (_serverResponseTwo.statusCode == 200) {
-      return parseData['key'];
-    } else {
-      throw JuntoException(
-        _serverResponse.reasonPhrase,
-        _serverResponse.statusCode,
+      // denote file type and get url, headers, and key of s3 bucket
+      final http.Response _serverResponse = await client.postWithoutEncoding(
+        _serverUrl,
+        body: {'content_type': fileType, 'content_length': contentLength},
       );
+      logger.logDebug(_serverResponse.body);
+      logger.logDebug(_serverResponse.statusCode.toString());
+
+      // parse response
+      final Map<String, dynamic> parseData =
+          JuntoHttp.handleResponse(_serverResponse);
+
+      // seralize into new headers
+      final Map<String, String> newHeaders = <String, String>{
+        'x-amz-acl': parseData['headers']['x-amz-acl'],
+        'x-amz-meta-user_id': parseData['headers']['x-amz-meta-user_id']
+      };
+
+      // turn file into bytes
+      final Uint8List fileAsBytes = file.readAsBytesSync();
+
+      // send put request to s3 bucket with url, new headers, and file as bytes
+      final http.Response _serverResponseTwo = await http.put(
+        parseData['signed_url'],
+        headers: newHeaders,
+        body: fileAsBytes,
+      );
+
+      // if successful, return the key for next steps
+      if (_serverResponseTwo.statusCode == 200) {
+        return parseData['key'];
+      } else {
+        throw JuntoException(
+          _serverResponse.reasonPhrase,
+          _serverResponse.statusCode,
+        );
+      }
+    } catch (e) {
+      logger.logException(e);
     }
   }
 

--- a/lib/backend/user_data_provider.dart
+++ b/lib/backend/user_data_provider.dart
@@ -53,21 +53,26 @@ class UserDataProvider extends ChangeNotifier {
   /// Update cached user information, called by [updateUser]
   Future<void> _setUserInformation(UserData user) async {
     final box = await Hive.openLazyBox(HiveBoxes.kAppBox, encryptionKey: key);
-    box.delete(HiveKeys.kUserData);
-    final userData = jsonEncode(user);
-    box.put(HiveKeys.kUserData, userData);
-    box.delete(HiveKeys.kUserId);
-    box.put(HiveKeys.kUserId, user.user.address);
+    await box.delete(HiveKeys.kUserData);
+    final userMap = user.toMap();
+    final userData = jsonEncode(userMap);
+    await box.put(HiveKeys.kUserData, userData);
+    await box.delete(HiveKeys.kUserId);
+    await box.put(HiveKeys.kUserId, user.user.address);
   }
 
   /// Updates the user information with [user]
   void updateUser(UserData user) {
-    assert(user.user.address == userAddress);
-    logger.logDebug(
-        'Current user address is equal to the updated user address: ${user.user.address == userAddress}');
-    _setUserInformation(user);
-    userProfile = user;
-    notifyListeners();
+    try {
+      assert(user.user.address == userAddress);
+      logger.logDebug(
+          'Current user address is equal to the updated user address: ${user.user.address == userAddress}');
+      _setUserInformation(user);
+      userProfile = user;
+      notifyListeners();
+    } catch (e) {
+      logger.logException(e);
+    }
   }
 
   Future<void> switchColumnLayout(ExpressionFeedLayout layout) async {

--- a/lib/screens/welcome/bloc/auth_bloc.dart
+++ b/lib/screens/welcome/bloc/auth_bloc.dart
@@ -54,9 +54,11 @@ class AuthBloc extends Bloc<AuthEvent, AuthState> {
       final user = await authRepo.registerUser(event.details);
       await userDataProvider.initialize();
       yield AuthenticatedState(user);
-    } on JuntoException catch (_) {
+    } on JuntoException catch (e) {
+      logger.logException(e);
       yield UnAuthenticatedState();
     } catch (error) {
+      logger.logException(error);
       yield UnAuthenticatedState();
     }
   }

--- a/lib/screens/welcome/bloc/auth_bloc.dart
+++ b/lib/screens/welcome/bloc/auth_bloc.dart
@@ -51,8 +51,10 @@ class AuthBloc extends Bloc<AuthEvent, AuthState> {
   Stream<AuthState> _mapSignUpEventState(SignUpEvent event) async* {
     yield LoadingState();
     try {
-      final user = await authRepo.registerUser(event.details);
+      final user =
+          await authRepo.registerUser(event.details, event.profilePicture);
       await userDataProvider.initialize();
+
       yield AuthenticatedState(user);
     } on JuntoException catch (e) {
       logger.logException(e);
@@ -60,6 +62,8 @@ class AuthBloc extends Bloc<AuthEvent, AuthState> {
     } catch (error) {
       logger.logException(error);
       yield UnAuthenticatedState();
+    } finally {
+      event?.completer?.complete();
     }
   }
 

--- a/lib/screens/welcome/bloc/auth_event.dart
+++ b/lib/screens/welcome/bloc/auth_event.dart
@@ -1,3 +1,6 @@
+import 'dart:async';
+import 'dart:io';
+
 import 'package:junto_beta_mobile/models/models.dart';
 import 'package:meta/meta.dart';
 
@@ -5,9 +8,12 @@ import 'package:meta/meta.dart';
 abstract class AuthEvent {}
 
 class SignUpEvent extends AuthEvent {
-  SignUpEvent(this.details);
+  SignUpEvent(this.details, this.profilePicture, this.completer);
 
   final UserAuthRegistrationDetails details;
+  final File profilePicture;
+  // this is a temporary solution until welcome.dart is rewritten to handle bloc
+  final Completer completer;
 }
 
 class LoginEvent extends AuthEvent {

--- a/lib/screens/welcome/sign_up_photos.dart
+++ b/lib/screens/welcome/sign_up_photos.dart
@@ -18,8 +18,8 @@ class SignUpPhotos extends StatefulWidget {
 class SignUpPhotosState extends State<SignUpPhotos> {
   File profilePictureOne;
 
-  List<File> returnDetails() {
-    return <File>[profilePictureOne];
+  File returnDetails() {
+    return profilePictureOne;
   }
 
   Widget _buildPhotoSelector({File profilePicture}) {

--- a/lib/screens/welcome/welcome.dart
+++ b/lib/screens/welcome/welcome.dart
@@ -1,3 +1,6 @@
+import 'dart:async';
+import 'dart:io';
+
 import 'package:flutter/material.dart';
 import 'package:flutter_bloc/flutter_bloc.dart';
 import 'package:hive/hive.dart';
@@ -49,7 +52,6 @@ class WelcomeState extends State<Welcome> {
   // Used when resetting password
   final ValueNotifier<String> _email = ValueNotifier("");
   String _currentTheme;
-  String _userAddress;
 
   PageController _welcomeController;
   PageController _signInController;
@@ -61,7 +63,7 @@ class WelcomeState extends State<Welcome> {
   String location;
   String gender;
   String website;
-  List<dynamic> profilePictures;
+  File profilePicture;
   String email;
   String password;
   String confirmPassword;
@@ -124,6 +126,7 @@ class WelcomeState extends State<Welcome> {
       email: email,
       name: name,
       password: password,
+      confirmPassword: confirmPassword,
       location: <String>[location],
       username: username,
       profileImage: <String>[],
@@ -136,62 +139,22 @@ class WelcomeState extends State<Welcome> {
 
     try {
       JuntoLoader.showLoader(context);
-      context.bloc<AuthBloc>().add(SignUpEvent(details));
-      _userAddress = userDataProvider.userAddress;
+      //TODO this is workaround because we need to wait for the registration to complete
+      final completer = Completer();
+      context
+          .bloc<AuthBloc>()
+          .add(SignUpEvent(details, profilePicture, completer));
+      await completer.future;
       JuntoLoader.hide();
     } catch (error) {
       JuntoLoader.hide();
-      print(error);
+      logger.logException(error);
       showDialog(
         context: context,
         builder: (BuildContext context) => SingleActionDialog(
-          dialogText: error.message,
+          dialogText: S.of(context).common_network_error,
         ),
       );
-    }
-
-    // Update user with profile photos
-    if (profilePictures[0] != null) {
-      // instantiate list to store photo keys retrieve from /s3
-      final List<String> _photoKeys = <String>[];
-      for (final dynamic image in profilePictures) {
-        if (image != null) {
-          try {
-            final String key =
-                await Provider.of<ExpressionRepo>(context, listen: false)
-                    .createPhoto(
-              false,
-              '.png',
-              image,
-            );
-            _photoKeys.add(key);
-          } catch (error) {
-            logger.logException(error);
-            JuntoLoader.hide();
-          }
-        }
-      }
-
-      Map<String, dynamic> _profilePictureKeys;
-
-      // instantiate data structure to update user with profile pictures
-      _profilePictureKeys = <String, dynamic>{
-        'profile_picture': <Map<String, dynamic>>[
-          <String, dynamic>{'index': 0, 'key': _photoKeys[0]},
-          if (_photoKeys.length == 2)
-            <String, dynamic>{'index': 1, 'key': _photoKeys[1]},
-        ]
-      };
-      // update user with profile photos
-      try {
-        await userRepository.updateUser(
-          profilePictures[0] == null ? _photoKeys : _profilePictureKeys,
-          _userAddress,
-        );
-      } catch (error) {
-        print(error);
-        JuntoLoader.hide();
-      }
     }
 
     JuntoLoader.hide();
@@ -427,8 +390,8 @@ class WelcomeState extends State<Welcome> {
         website = _aboutPageModel.website;
         print(bio.length);
       } else if (_currentIndex == 5) {
-        profilePictures = signUpPhotosKey.currentState.returnDetails();
-        print(profilePictures);
+        profilePicture = signUpPhotosKey.currentState.returnDetails();
+        print(profilePicture);
       } else if (_currentIndex == 6) {
         email = signUpRegisterKey.currentState.returnDetails()['email'];
         password = signUpRegisterKey.currentState.returnDetails()['password'];
@@ -450,7 +413,8 @@ class WelcomeState extends State<Welcome> {
         }
         JuntoLoader.showLoader(context, color: Colors.transparent);
         // verify email address
-        await authRepo.verifyEmail(email);
+        final result = await authRepo.verifyEmail(email);
+        logger.logDebug(result);
         JuntoLoader.hide();
       }
       // transition to next page of sign up flow

--- a/lib/screens/welcome/welcome.dart
+++ b/lib/screens/welcome/welcome.dart
@@ -2,6 +2,7 @@ import 'package:flutter/material.dart';
 import 'package:flutter_bloc/flutter_bloc.dart';
 import 'package:hive/hive.dart';
 import 'package:junto_beta_mobile/api.dart';
+import 'package:junto_beta_mobile/app/logger/logger.dart';
 import 'package:junto_beta_mobile/backend/backend.dart';
 import 'package:junto_beta_mobile/generated/l10n.dart';
 import 'package:junto_beta_mobile/hive_keys.dart';
@@ -77,6 +78,7 @@ class WelcomeState extends State<Welcome> {
 
   UserRepo userRepository;
   UserDataProvider userDataProvider;
+  AuthRepo authRepo;
 
   @override
   void initState() {
@@ -103,6 +105,7 @@ class WelcomeState extends State<Welcome> {
     //TODO: don't perform user signup in state of welcome.dart
     userRepository = Provider.of<UserRepo>(context, listen: false);
     userDataProvider = Provider.of<UserDataProvider>(context, listen: false);
+    authRepo = Provider.of<AuthRepo>(context, listen: false);
   }
 
   @override
@@ -163,7 +166,7 @@ class WelcomeState extends State<Welcome> {
             );
             _photoKeys.add(key);
           } catch (error) {
-            print(error);
+            logger.logException(error);
             JuntoLoader.hide();
           }
         }
@@ -401,8 +404,7 @@ class WelcomeState extends State<Welcome> {
           JuntoLoader.showLoader(context, color: Colors.transparent);
           // ensure username is not taken or reserved
           final Map<String, dynamic> validateUserResponse =
-              await Provider.of<AuthRepo>(context, listen: false)
-                  .validateUser(username: username);
+              await authRepo.validateUser(username: username);
           JuntoLoader.hide();
           final bool usernameIsAvailable =
               validateUserResponse['valid_username'];
@@ -448,7 +450,7 @@ class WelcomeState extends State<Welcome> {
         }
         JuntoLoader.showLoader(context, color: Colors.transparent);
         // verify email address
-        await Provider.of<AuthRepo>(context, listen: false).verifyEmail(email);
+        await authRepo.verifyEmail(email);
         JuntoLoader.hide();
       }
       // transition to next page of sign up flow

